*** ./s02/TimerQueue.cpp	2015-12-12 17:06:22.964000000 +0800
--- ./s03/TimerQueue.cpp	2015-12-12 22:07:39.604000000 +0800
***************
*** 88,100 ****
                               Timestamp when,
                               double interval){
      Timer* timer = new Timer(cb, when, interval);
      loop_->assertInLoopThread();
      bool earliestChanged = insert(timer);
  
      if(earliestChanged){
          resetTimerfd(timerfd_, timer->expiration());
      }
-     return TimerId(timer);
  }
  
  void TimerQueue::handleRead(){
--- 88,105 ----
                               Timestamp when,
                               double interval){
      Timer* timer = new Timer(cb, when, interval);
+     loop_->runInLoop(
+             boost::bind(&TimerQueue::addTimerInLoop, this, timer));
+     return TimerId(timer);
+ }
+ 
+ void TimerQueue::addTimerInLoop(Timer* timer){
      loop_->assertInLoopThread();
      bool earliestChanged = insert(timer);
  
      if(earliestChanged){
          resetTimerfd(timerfd_, timer->expiration());
      }
  }
  
  void TimerQueue::handleRead(){
