*** ./s04/SocketsOps.cpp	2015-12-16 17:21:54.790340582 +0800
--- ./s05/SocketsOps.cpp	2015-12-16 22:39:24.388000000 +0800
***************
*** 140,142 ****
--- 140,152 ----
          LOG_SYSERR << "sockets::fromHostPort";
      }
  }
+ 
+ struct sockaddr_in sockets::getLocalAddr(int sockfd){
+     struct sockaddr_in localaddr;
+     bzero(&localaddr, sizeof(localaddr));
+     socklen_t addrlen = sizeof(localaddr);
+     if(::getsockname(sockfd, sockaddr_cast(&localaddr), &addrlen) < 0){
+         LOG_SYSERR << "sockets::getLocalAddr";
+     }
+     return localaddr;
+ }
