*** ./s05/SocketsOps.cpp	2015-12-16 22:39:24.388000000 +0800
--- ./s06/SocketsOps.cpp	2015-12-23 17:18:42.079373000 +0800
***************
*** 150,152 ****
--- 150,163 ----
      }
      return localaddr;
  }
+ 
+ int sockets::getSocketError(int sockfd){
+     int optval;
+     socklen_t optlen = sizeof(optval);
+ 
+     if(::getsockopt(sockfd, SOL_SOCKET, SO_ERROR, &optval, &optlen) < 0){
+         return errno;
+     }else{
+         return optval;
+     }
+ }
