*** ./s05/Callbacks.h	2015-12-16 22:07:43.084150000 +0800
--- ./s06/Callbacks.h	2015-12-22 21:38:27.420000000 +0800
***************
*** 18,23 ****
--- 18,24 ----
      typedef boost::function<void (const TcpConnectionPtr&,
                                    const char* data,
                                    ssize_t len)> MessageCallback;
+     typedef boost::function<void (const TcpConnectionPtr&)> CloseCallback;
  
  }
  
