*** ./s06/Callbacks.h	2015-12-22 21:38:27.420000000 +0800
--- ./s07/Callbacks.h	2015-12-27 16:27:44.332000000 +0800
***************
*** 10,23 ****
      
      // All client visible callbacks go here.
  
      class TcpConnection;
      typedef boost::shared_ptr<TcpConnection> TcpConnectionPtr;
  
      typedef boost::function<void()> TimerCallback;
      typedef boost::function<void (const TcpConnectionPtr&)> ConnectionCallback;
      typedef boost::function<void (const TcpConnectionPtr&,
!                                   const char* data,
!                                   ssize_t len)> MessageCallback;
      typedef boost::function<void (const TcpConnectionPtr&)> CloseCallback;
  
  }
--- 10,24 ----
      
      // All client visible callbacks go here.
  
+     class Buffer;
      class TcpConnection;
      typedef boost::shared_ptr<TcpConnection> TcpConnectionPtr;
  
      typedef boost::function<void()> TimerCallback;
      typedef boost::function<void (const TcpConnectionPtr&)> ConnectionCallback;
      typedef boost::function<void (const TcpConnectionPtr&,
!                                   Buffer* buf,
!                                   Timestamp)> MessageCallback;
      typedef boost::function<void (const TcpConnectionPtr&)> CloseCallback;
  
  }
