*** ./s06/TcpConnection.h	2015-12-23 17:22:30.991373000 +0800
--- ./s07/TcpConnection.h	2015-12-27 16:40:43.120000000 +0800
***************
*** 1,6 ****
--- 1,7 ----
  #ifndef MUDUO_TCPCONNECTION_H
  #define MUDUO_TCPCONNECTION_H
  
+ #include "Buffer.h"
  #include "Callbacks.h"
  #include "InetAddress.h"
  
***************
*** 80,86 ****
              }
  
              // 把readable事件交给messagecallback
!             void handleRead();
              void handleWrite();
              void handleClose();
              void handleError();
--- 81,87 ----
              }
  
              // 把readable事件交给messagecallback
!             void handleRead(Timestamp receiveTime);
              void handleWrite();
              void handleClose();
              void handleError();
***************
*** 96,101 ****
--- 97,103 ----
              ConnectionCallback connectionCallback_;
              MessageCallback messageCallback_;
              CloseCallback closeCallback_;
+             Buffer inputBuffer_;
      };
  
      typedef boost::shared_ptr<TcpConnection> TcpConnectionPtr;
