*** ./s07/SocketsOps.h	2015-12-27 16:19:05.876000000 +0800
--- ./s08/SocketsOps.h	2015-12-28 15:44:08.143851000 +0800
***************
*** 40,45 ****
--- 40,46 ----
      void listenOrDie(int sockfd);
      int accept(int sockfd, struct sockaddr_in* addr);
      void close(int sockfd);
+     void shutdownWrite(int sockfd);
  
      void toHostPort(char* buf, size_t size,
                      const struct sockaddr_in& addr);
