*** ./s07/SocketsOps.cpp	2015-12-27 16:19:05.876000000 +0800
--- ./s08/SocketsOps.cpp	2015-12-28 15:58:44.032000000 +0800
***************
*** 124,129 ****
--- 124,135 ----
      }
  }
  
+ void sockets::shutdownWrite(int sockfd){
+     if(::shutdown(sockfd, SHUT_WR) < 0){
+         LOG_SYSERR << "sockets::shutdownWrite";
+     }
+ }
+ 
  void sockets::toHostPort(char* buf, size_t size, 
                           const struct sockaddr_in& addr){
      char host[INET_ADDRSTRLEN] = "INVALID";
