*** ./s08/Callbacks.h	2015-12-28 15:33:56.663851000 +0800
--- ./s09/Callbacks.h	2015-12-29 15:37:59.212000000 +0800
***************
*** 19,24 ****
--- 19,25 ----
      typedef boost::function<void (const TcpConnectionPtr&,
                                    Buffer* buf,
                                    Timestamp)> MessageCallback;
+     typedef boost::function<void (const TcpConnectionPtr&)> WriteCompleteCallback;
      typedef boost::function<void (const TcpConnectionPtr&)> CloseCallback;
  
  }
