*** ./s10/SocketsOps.h	2015-12-30 21:04:32.296000000 +0800
--- ./s11/SocketsOps.h	2015-12-31 16:09:51.492000000 +0800
***************
*** 36,41 ****
--- 36,42 ----
      /// abort if any error
      int createNonblockingOrDie();
  
+     int connect(int sockfd, const struct sockaddr_in& addr);
      void bindOrDie(int sockfd, const struct sockaddr_in& addr);
      void listenOrDie(int sockfd);
      int accept(int sockfd, struct sockaddr_in* addr);
***************
*** 48,57 ****
                      struct sockaddr_in* addr);
  
      struct sockaddr_in getLocalAddr(int sockfd);
  
      int getSocketError(int sockfd);
  
  }
  }
  
! #endif
--- 49,60 ----
                      struct sockaddr_in* addr);
  
      struct sockaddr_in getLocalAddr(int sockfd);
+     struct sockaddr_in getPeerAddr(int sockfd);
  
      int getSocketError(int sockfd);
+     bool isSelfConnect(int sockfd);
  
  }
  }
  
! #endif // MUDUO_NET_SOCKETSOPS_H
